tools/please_cc/please_cc.go

```diff
@@ -0,0 +1,173 @@
+// Binary please_cc implements a tool for invoking C/C++ compilers and linkers using command line options that are
+// dynamically selected depending on the compiler or linker being invoked.
+//
+// please_cc operates in one of two modes, indicated by the value of the first argument passed to it:
+//
+// - `cc` ("compiler mode") invokes the compiler - and, implicitly, the linker - in a C or C++ toolchain.
+// - `ld` ("linker mode") invokes a linker independently of a C or C++ toolchain.
+//
+// The second argument identifies the (relative or absolute) path to the compiler or linker. If the path is relative,
+// please_cc follows the usual conventions for converting relative paths into absolute paths before invoking the
+// compiler or linker.
+//
+// The remaining arguments are command line options to be passed through to the compiler or linker. Arguments surrounded
+// by `{{ ` and ` }}` are evaluated as expressions (see [github.com/please-build/cc-rules/tools/please_cc/expr]) after
+// those delimiters have been stripped; the result of each evaluation is substituted into the list of command line
+// options passed to the compiler or linker. The identity of the compiler and/or linker is described in the environment
+// in which the expression is evaluated, which allows for different command line options to be passed through for
+// different compilers and/or linkers.
+//
+// For example:
+//
+// - Only make warnings fatal when compiling with GCC:
+//   `please_cc cc -o example '{{ gcc ? "-Werror" }}' example.c`
+// - Compile C++ modules with GCC and Clang <= 15 with the `-fmodules-ts` option, or Clang >= 16 with the (equivalent,
+//   non-deprecated) `-std=c++20` option:
+//   `please_cc cc -o example '{{ gcc || (clang && clang <= 15) ? "-fmodules-ts" : "-std=c++20" }}' example.cc`
+// - Link objects using the old ld64 code path if linking with Apple's new ld linker (enabled using `-ld64` prior to
+//   Xcode 15.1, and `-ld_classic` from Xcode 15.1 onwards):
+//   `please_cc ld '{{ appleld ? (appleld >= 1022.1 ? "-ld_classic" : "-ld64") }}' obj1.o obj2.o -o example
+//
+// please_cc is known to be compatible with:
+//
+// - GCC >= 9
```

Comment:
Add the usable variables to this list as well?

Reply:
I'll put this in a table.